Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,625 @@
+      apply le_trans (mul_le_mul_of_nonneg_right
+        (rpow_le_rpow (smoothingFun_nonneg μ hμ1 _) h a_in.1)
+        (rpow_nonneg (smoothingFun_nonneg μ hμ1 _) _))
+      rw [hb, ← rpow_add_of_nonneg (smoothingFun_nonneg μ hμ1 _) a_in.1
+        (sub_nonneg.mpr a_in.2), add_sub, add_sub_cancel_left, rpow_one]
+    · rw [add_le_add_iff_right]
+      apply le_trans (mul_le_mul_of_nonneg_left
+        (rpow_le_rpow (smoothingFun_nonneg μ hμ1 _) (le_of_lt (not_le.mp h)) b_in.1)
+        (rpow_nonneg (smoothingFun_nonneg μ hμ1 _) _))
+      rw [hb, ← rpow_add_of_nonneg (smoothingFun_nonneg μ hμ1 _) a_in.1
+        (sub_nonneg.mpr a_in.2), add_sub, add_sub_cancel_left, rpow_one]
+  apply le_trans _ h_mul
+  have hex : ∃ n : PNat, μ (x ^ mu (ψ n)) ^ (1 / (ψ n : ℝ)) * μ (y ^ nu (ψ n)) ^ (1 / (ψ n : ℝ)) <
+      smoothingFun μ x ^ a * smoothingFun μ y ^ b + ε :=
+    Filter.exists_lt_of_limsup_le (bddAbove_range_mul (μ_bddAbove μ hμ1 hmu_le _ _)
+      (fun n => rpow_nonneg (apply_nonneg _ _) _) (μ_bddAbove μ hμ1 hnu_le _ _)
+        fun n => rpow_nonneg (apply_nonneg _ _) _).isBoundedUnder_of_range hxy hε
+  obtain ⟨N, hN⟩ := hex
+  apply le_trans (ciInf_le (smoothingSeminormSeq_bddBelow μ _)
+    ⟨ψ N, lt_of_le_of_lt (_root_.zero_le (ψ 0)) (hψ_mono.lt_iff_lt.mpr N.pos)⟩)
+  apply le_trans _ hN.le
+  simp only [PNat.mk_coe, hnu, ← mul_rpow (apply_nonneg μ _) (apply_nonneg μ _)]
+  exact mu_property μ hn (ψ N)
+
+end IsNonarchimedean
+
+/-- If `μ 1 ≤ 1` and `μ` is nonarchimedean, then `smoothingFun` is a ring seminorm. -/
+def smoothingSeminorm (hμ1 : μ 1 ≤ 1) (hna : IsNonarchimedean μ) : RingSeminorm R where
+  toFun     := smoothingFun μ
+  map_zero' := by
+    apply tendsto_nhds_unique_of_eventuallyEq (tendsto_smoothingFun_of_map_one_le_one μ hμ1 0)
+      tendsto_const_nhds
+    simp only [EventuallyEq, eventually_atTop, ge_iff_le]
+    use 1
+    intro n hn
+    simp only [smoothingSeminormSeq]
+    rw [zero_pow (pos_iff_ne_zero.mp hn), map_zero, zero_rpow]
+    exact one_div_ne_zero (cast_ne_zero.mpr (one_le_iff_ne_zero.mp hn))
+  add_le' _ _ :=  (isNonarchimedean_smoothingFun μ hμ1 hna).add_le (smoothingFun_nonneg μ hμ1)
+  neg' n := by
+    simp only [smoothingFun]
+    congr
+    ext n
+    rw [neg_pow]
+    rcases neg_one_pow_eq_or R n with hpos | hneg
+    · rw [hpos, one_mul]
+    · rw [hneg, neg_one_mul, map_neg_eq_map μ]
+  mul_le' x y := by
+    apply le_of_tendsto_of_tendsto' (tendsto_smoothingFun_of_map_one_le_one μ hμ1 (x * y))
+      (Tendsto.mul (tendsto_smoothingFun_of_map_one_le_one μ hμ1 x)
+        (tendsto_smoothingFun_of_map_one_le_one μ hμ1 y))
+    intro n
+    have hn : 0 ≤ 1 / (n : ℝ) := by simp only [one_div, inv_nonneg, cast_nonneg]
+    simp only [smoothingSeminormSeq]
+    rw [← mul_rpow (apply_nonneg μ _) (apply_nonneg μ _), mul_pow]
+    exact rpow_le_rpow (apply_nonneg μ _) (map_mul_le_mul μ _ _) hn
+
+/-- If `μ 1 ≤ 1` and `μ` is nonarchimedean, then `smoothingSeminorm μ hμ1 hna 1 ≤ 1`. -/
```

Comment:
```suggestion
/-- If `μ 1 ≤ 1` and `μ` is nonarchimedean, then `smoothingSeminorm μ 1 ≤ 1`. -/
```
Although this does not reflect the code verbatim, I think it is more readable and useful.